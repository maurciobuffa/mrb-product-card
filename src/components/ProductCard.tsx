import React, { createContext, CSSProperties } from "react";
import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import {
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from "../interfaces/interfaces";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface ProductCardProps {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, isMaxCountReached, reset } =
    useProduct({
      product,
      onChange,
      value,
      initialValues,
    });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          increaseBy,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
