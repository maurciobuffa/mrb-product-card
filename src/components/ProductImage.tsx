import React, { CSSProperties, useContext } from "react";
import styles from "../styles/styles.module.css";
import noImage from "../assets/no-image.jpg";
import { ProductContext } from "./ProductCard";

export interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) => {
  const { product } = useContext(ProductContext);

  const imageToShow = img || product.img || noImage;

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imageToShow}
      alt="Product"
      style={style}
    />
  );
};
