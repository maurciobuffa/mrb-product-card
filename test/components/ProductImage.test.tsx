import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components'
import { product2 } from '../data/products'

describe('ProductImage', () => {
  test('renders custom Image component', () => {
    const wrapper = renderer.create(<ProductImage img="https://hi.jpg" className="custom-class" />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  test('should render component with product image', () => {
    const wrapper = renderer.create(<ProductCard product={product2}>
      {
        () => (
          <ProductImage />
        )
      }
    </ProductCard>)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
