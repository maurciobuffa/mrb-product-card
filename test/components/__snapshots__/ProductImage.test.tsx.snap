// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage renders custom Image component 1`] = `
<img
  alt="Product"
  className="productImg custom-class"
  src="https://hi.jpg"
/>
`;

exports[`ProductImage should render component with product image 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
